src/flb_plugin_proxy.c

```diff
@@ -68,43 +70,113 @@ static void proxy_cb_flush(struct flb_event_chunk *event_chunk,
     FLB_OUTPUT_RETURN(ret);
 }
 
-static void proxy_cb_in_thread_callback(int write_fd, void *data)
+static int flb_proxy_input_cb_timer(struct flb_input_instance *ins,
+                                    struct flb_config *config, void *in_context)
 {
     int ret = FLB_OK;
     size_t len = 0;
-    struct flb_input_thread *it = data;
-    mpack_writer_t *writer = &it->writer;
-    struct flb_plugin_input_proxy_thread_config *ctx;
+    void *data = NULL;
+    struct flb_plugin_input_proxy_context *ctx = (struct flb_plugin_input_proxy_context *) in_context;
 
 #ifdef FLB_HAVE_PROXY_GO
-    ctx = container_of(it, struct flb_plugin_input_proxy_thread_config, it);
-
     if (ctx->proxy->def->proxy == FLB_PROXY_GOLANG) {
-        while(!flb_input_thread_exited(it)) {
-            flb_trace("[GO] entering go_collect()");
-            ret = proxy_go_input_collect(ctx->proxy, &it->data, &len);
-
-            if (ret == -1) {
-                flb_errno();
-                break;
-            }
-
-            mpack_write_object_bytes(writer, it->data, len);
-            mpack_writer_flush_message(writer);
-            fflush(it->write_file);
-
-            if (!it->data) {
-                free(data);
-            }
-
-            ret = proxy_go_input_cleanup(ctx->proxy, it->data);
-            if (ret == -1) {
-                flb_errno();
-                break;
-            }
+        flb_trace("[GO] entering go_collect()");
+        ret = proxy_go_input_collect(ctx->proxy, &data, &len);
+
+        if (ret == -1) {
+            flb_errno();
+            FLB_INPUT_RETURN(1);
+        }
+
+        flb_input_chunk_append_raw(ins, NULL, 0, data, len);
+
+        if (!data) {
+            free(data);
+        }
+
+        ret = proxy_go_input_cleanup(ctx->proxy, data);
+        if (ret == -1) {
+            flb_errno();
+            FLB_INPUT_RETURN(1);
         }
     }
 #endif
+
+    FLB_INPUT_RETURN(0);
+}
+
+static int flb_proxy_input_cb_init(struct flb_input_instance *ins,
+                                   struct flb_config *config, void *data)
+{
+    int ret = -1;
+    struct flb_plugin_input_proxy_context *ctx;
+    struct flb_plugin_proxy_context *pc;
+
+    /* Allocate space for the configuration context */
+    ctx = flb_malloc(sizeof(struct flb_plugin_input_proxy_context));
+    if (!ctx) {
+        flb_errno();
+        return -1;
+    }
+
+    /* Before to initialize for proxy, set the proxy instance reference */
+    pc = (struct flb_plugin_proxy_context *)(ins->context);
+    ctx->proxy = pc->proxy;
+
+    /* Before to initialize, set the instance reference */
+    pc->proxy->instance = ins;
+
+    /* Based on 'proxy', use the proper handler */
+    if (pc->proxy->def->proxy == FLB_PROXY_GOLANG) {
+#ifdef FLB_HAVE_PROXY_GO
```

Comment:
Shouldn't we return an error in this code path if `FLB_HAVE_PROXY_GO` is undefined?

Reply:
I added a else clause to raise error logat https://github.com/fluent/fluent-bit/pull/5056/commits/37bb521fd44839b6f0b1a3a7c6b1fe2630f7adbe.